Fix, clarify, and simplify content type schemas (#2351)
OAI/OpenAPI-Specification: versions/3.1.0.md

##### Considerations for File Uploads

> In contrast with the 2.0 specification, `file` input/output content in OpenAPI is described with the same semantics as any other schema type.  In contrast with the 3.0 specification, such schemas either omit the `type` (in place of `format: binary`), or use `contentMediaType` and `contentEncoding` with `type: string`. The `contentEncoding` keyword supports all encodings defined in [RFC4648](https://tools.ietf.org/html/rfc4648), including "base64" (which replaces `format: byte`) and "base64url", as well as "quoted-printable" from [RFC2045](https://tools.ietf.org/html/rfc2045#section-6.7).

Reviewer: Here is my proposed rewording to clarify the priority of `Content-Encoding` headers (in various places) and the JSON Schema `contentEncoding`:
```suggestion
In contrast with the 2.0 specification, `file` input/output content in OpenAPI is described with the same semantics as any other schema type. 

In contrast with the 3.0 specification, the `format` keyword has no effect on the content-encoding of the schema. The Content-Encoding for the schema can be specified explicitly with the `Content-Encoding` parameter of the operation, for schemas in request bodies, or in a `Content-Encoding` header in the `headers` of the response, for response schemas, or in a `Content-Encoding` header in the `headers` of an Encoding object associated with a request or response body property. JSON Schema also offers a `contentEncoding` keyword, which may be used to specify the `Content-Encoding` for the schema, but `contentEncoding` SHALL be ignored if a `Content-Encoding` header is defined. The `contentEncoding` keyword supports all encodings defined in [RFC4648](https://tools.ietf.org/html/rfc4648), including "base64" (which replaces `format: byte`) and "base64url", as well as "quoted-printable" from [RFC2045](https://tools.ietf.org/html/rfc2045#section-6.7). When not explicitly specified, the content-encoding

JSON Schema also offers a `contentMediaType` keyword.  However, when the media type is already specified by the Media Type Object's key, or by the `contentType` field of an [Encoding Object](#encodingObject), the `contentMediaType` keyword SHALL be ignored if present.
```

Author: @mkistler the `Content-Encoding` HTTP header and `contentEncoding` JSON Schema keyword are unrelated.  The HTTP header is for things like gzip that the HTTP client and server will encode/decode automatically.  The JSON Schema keyword is for when both HTTP and the media type think that something is text, but the application needs to be informed that some specific piece of text needs to be encoded/decoded in a particular way.

Author: You could in fact have a gzipped payload that includes a base64-encoded PNG file.  That would use both the header and the keyword.

Reviewer: @handrews I struggling to understand this all.  Do you mean to say that `contentEncoding` keyword has no effect whatsoever on the value that should/will be passed in the `Content-Encoding` header? If that is the case then it certainly will confuse some readers (like me :smile:).

In the example you give, how would this look in the API doc?  Maybe:
```
  /v1/photos:
    post:
      parameters:
        - name: Content-Encoding
           in: header
           schema:
             type: string
             enum:
               - gzip
        requestBody:
          content:
            image/png:
              schema:
                contentEncoding: base64
```

And the http request would specify:
- `Content-type: image/png`
- `Content-encoding: gzip`
- a gzipped, base64-encoded image/png

Is that right?

If I've got this right, then I think you are saying that _both_ the encoding specified in the `Content-Encoding` header _and_ the encoding specified in the `contentEncoding` keyword are applied to the request body (or whichever element they pertain to).

If that's right, I'll make a stab at rewording my suggestion to describe this.

Author: @mkistler yeah this is the challenge of combining multiple standards that were made in different contexts at different times.

If you go back to MIME, there is a `Content-Transfer-Encoding` header, which is where values such as `base64` from RFC 4648 (or its predecessors) are used.  Perhaps we should have used `mediaEncoding` or something like that in JSON Schema (the naming of that keyword has a slightly convoluted history that partially predates me), but we didn't.  I might have only learned about HTTP `Content-Encoding` while researching this PR, come to think of it.

I believe that what you have for the HTTP request is correct.  At this stage with the modifications y'all have made recently, I'm a little unclear on where we ended up on `type` being absent vs `"type": "string"`, but what you have (with `type` absent) is what I would have written.  And is compatible with the latest JSON Schema draft which notes that an application MAY apply a JSON Schema to other media types, including binary media types, and notes that many keywords have no sensible meaning with binary types.

> In contrast with the 2.0 specification, `file` input/output content in OpenAPI is described with the same semantics as any other schema type.  In contrast with the 3.0 specification, such schemas either omit the `type` (in place of `format: binary`), or use `contentMediaType` and `contentEncoding` with `type: string`. The `contentEncoding` keyword supports all encodings defined in [RFC4648](https://tools.ietf.org/html/rfc4648), including "base64" (which replaces `format: byte`) and "base64url", as well as "quoted-printable" from [RFC2045](https://tools.ietf.org/html/rfc2045#section-6.7).

Reviewer: Second attempt:

```suggestion
In contrast with the 2.0 specification, `file` input/output content in OpenAPI is described with the same semantics as any other schema type.

In contrast with the 3.0 specification, the `format` keyword has no effect on the content-encoding of the schema. JSON Schema offers a `contentEncoding` keyword, which may be used to specify the `Content-Encoding` for the schema. The `contentEncoding` keyword supports all encodings defined in [RFC4648](https://tools.ietf.org/html/rfc4648), including "base64" and "base64url", as well as "quoted-printable" from [RFC2045](https://tools.ietf.org/html/rfc2045#section-6.7). The encoding specified by the `contentEncoding` keyword is independent of an encoding specified by the `Content-Type` header in the request or response or metadata of a multipart body -- when both are present, the encoding specified in the `contentEncoding` is applied first and then the encoding specified in the `Content-Type` header.

JSON Schema also offers a `contentMediaType` keyword.  However, when the media type is already specified by the Media Type Object's key, or by the `contentType` field of an [Encoding Object](#encodingObject), the `contentMediaType` keyword SHALL be ignored if present.
```

```yaml
content:

>     text/plain:

Reviewer 2: I believe this should stay as image/png to convey the semantics of the HTTP payload.

Author: @darrelmiller I don't remember why I did that, but I'll take your word for it.  I am having a difficult time finding a reference (which I was hoping would remind me why I thought this was correct).  But yeah, the JSON Schema provides all of the processing information and the logical output is `image/png` so that makes sense.  And of course the text that asserted this should indeed be removed.  Lemme do that right now...

        schema:
            type: string
            contentMediaType: image/png
            contentEncoding: base64
```

> Note that the `Content-Type` is `text/plain`, since from an HTTP perspective, base64-encoded data is simply text.  The `contentMediaType` and `contentEncoding` fields inform the application how to encode and decode the resource from and to its actual media type.

Reviewer 2: Believe this text should be removed.

Author: @darrelmiller I have changed this paragraph to explain why there is a redundancy, but if you feel the new text is more confusing than helpful, I'm happy to remove it.

* If the property is a primitive, or an array of primitive values, the default Content-Type is `text/plain`
* If the property is complex, or an array of complex values, the default Content-Type is `application/json`

> * If the property is a `type: string` with a `contentEncoding`, the default Content-Type is `text/plain`, and the media type of the embedded resource is specified in `contentMediaType`

Reviewer 2: ```suggestion
* If the property is a `type: string` with a `contentEncoding`, the default Content-Type is `application/octet-stream`
```
Changed this back to match wording here: https://tools.ietf.org/html/draft-bhutton-json-schema-validation-00#section-8.3

##### Fixed Fields
Field Name | Type | Description
---|:---:|---

> <a name="encodingContentType"></a>contentType | `string` | The Content-Type for encoding a specific property. Default value depends on the property type: when `type` is absent - `application/octet-stream`; for primitive types - `text/plain`; for `object` - `application/json`; when `type` is `string` and `contentEncoding` is present, the default Content-Type is `text/plain`, and the media type of the encoded resource is specified in `contentMediaType`; for `array` – the default is defined based on the inner type. The value can be a specific media type (e.g. `application/json`), a wildcard media type (e.g. `image/*`), or a comma-separated list of the two types.

Reviewer: I thought that we agreed in the 1/28 TSC meeting that when `type` is `string` the default Content-Type should be `application/octet-stream`.

> <a name="encodingContentType"></a>contentType | `string` | The Content-Type for encoding a specific property. Default value depends on the property type: when `type` is absent - `application/octet-stream`; for primitive types - `text/plain`; for `object` - `application/json`; when `type` is `string` and `contentEncoding` is present, the default Content-Type is `text/plain`, and the media type of the encoded resource is specified in `contentMediaType`; for `array` – the default is defined based on the inner type. The value can be a specific media type (e.g. `application/json`), a wildcard media type (e.g. `image/*`), or a comma-separated list of the two types.

Reviewer 2: ```suggestion
<a name="encodingContentType"></a>contentType | `string` | The Content-Type for encoding a specific property. Default value depends on the property type: for `object` - `application/json`;  for `array` – the default is defined based on the inner type; for all other cases the default is `application/octet-stream`. The value can be a specific media type (e.g. `application/json`), a wildcard media type (e.g. `image/*`), or a comma-separated list of the two types. 
```

Author: @darrelmiller I'm fairly sure we went over the wording and cases for this in a call and there's a reason it was done this way.  I confess I can't remember what that was.
